Add absolute values and convert to Minion

niklasdewally opened the pull request:
Add support for absolute values, in particular:

* Parsing from JSON.
* The `Abs` expression variant.
* The `FlatAbsEq` Minion constraint.
* Flattening and conversion to Minion.

ozgurakgun commented on crates/conjure_core/src/rules/minion.rs:
```diff
@@ -405,6 +405,36 @@ fn introduce_modeq(expr: &Expr, _: &Model) -> ApplicationResult {
+fn introduce_abseq(expr: &Expr, _: &Model) -> ApplicationResult {
+    let (x, abs_y): (Atom, Expr) = match expr.clone() {
+        Expr::Eq(_, a, b) => {
+            let a_atom: Option<&Atom> = (&*a).try_into().ok();
```
really looking forward to [box_patterns](https://github.com/rust-lang/rust/issues/29641) to avoid this kind of thing

ozgurakgun replied:
oh, apparently box_patterns is no more, it's going to be [deref patterns](https://github.com/rust-lang/rust/issues/87121) instead.

ozgurakgun commented on crates/conjure_core/src/ast/expressions.rs:
```diff
@@ -370,8 +385,9 @@ impl Expression {
-            // _ => bug!("Cannot calculate domain of {:?}", self),
+            Expression::Abs(_, a) => a
+                .domain_of(vars)?
+                .apply_i32(|a, _| Some(a.abs()), &a.domain_of(vars)?),
```
so does this apply abs to all ints in the domain? what would happen to something like int(-7..-2)? it should be int(2..7).

also int(-2..3) could be turned into int(2..3) if we are not careful, though it should be int(0..3)

alternatively we can convert int(-2..3) to int(-2,-1,0,1,2,3), then map abs, sort, deduplicate, then we would end up with  int(0,1,2,3). we can optionally convert this back to int(0..3)

this would ve very inefficient with huge domains though, so not ideal.

niklasdewally replied:
> so does this apply abs to all ints in the domain? what would happen to something like int(-7..-2)? it should be int(2..7).

IIrc, this function deals with the bounds flipping internally.
We already use it for numerous other operations where the bounds can flip.

ozgurakgun replied:
How about the second part? Can you add a test case with -2..3 in it to see what happens please?

niklasdewally commented:
@ozgurakgun I have added a bug-fix for division domains to fix abs-03 test to this commit.

Updated commit message: 

> As part of this change, a division domain bug is fixed. In our domain
> calculation we use truncating division, the default Rust integer
> division operator. However, the semantics of Essence Prime specify floor
> division.
> 
> Truncating division has the effect of rounding towards 0.
> 
> Truncating division does `-5 / 2 =-2` and `5/2 = 2`. However, we want
> `-5/2 = -3` and `5/2 = 2`.
> 
> This caused the domains of variables containing negative divisions to
> sometimes be one too high. This caused abs-03 to give different
> solutions to Conjure.

ozgurakgun commented on conjure_oxide/tests/integration/basic/abs/01-simple/input.eprime:
```diff
@@ -0,0 +1,5 @@
+language ESSENCE' 1.0
+
+find x,y: int(-5..5)
```
change this to something other that -5..5 and we would be testing the bound flipping?

ozgurakgun commented on conjure_oxide/tests/integration/basic/abs/03-nested/input.eprime:
```diff
@@ -0,0 +1,7 @@
+language ESSENCE' 1.0
+
+find x,y,z: int(-5..2)
```
this might already be testing the -2..3 case I was worried about, but still adding another one would make me happier. just `find x : int(-2..3) such that |x| = 1` would be enough.

niklasdewally replied:
Apart from an unrelated typo, this test works

ozgurakgun commented on conjure_oxide/tests/integration/basic/abs/04-bounds/input.eprime:
```diff
@@ -0,0 +1,7 @@
+find x : int(-2..3) 
+
+such that 
+|x| = 1 
```
ok, so this is not enough since |x| is not extracted into an aux var, of course.

would something like `y * |x| = 2` with `find y : int(1..3)` create an aux for `|x|`?

niklasdewally replied:
Works 👍🏻
